zkevm-circuits/src/tx_circuit/sign_verify.rs

```diff
@@ -0,0 +1,992 @@
+        offset: usize,
+        is_enabled: F,
+        input_rlc: F,
+        input_len: usize,
+        output_rlc: F,
+    ) -> Result<(), Error> {
+        for (name, column, value) in &[
+            ("is_enabled", self.keccak_table[0], is_enabled),
+            ("input_rlc", self.keccak_table[1], input_rlc),
+            ("input_len", self.keccak_table[2], F::from(input_len as u64)),
+            ("output_rlc", self.keccak_table[3], output_rlc),
+        ] {
+            region.assign_advice(
+                || format!("Keccak table assign {} {}", name, offset),
+                *column,
+                offset,
+                || Ok(*value),
+            )?;
+        }
+        Ok(())
+    }
+
+    pub fn load_keccak(
+        &self,
+        layouter: &mut impl Layouter<F>,
+        auxs: Vec<KeccakAux>,
+        randomness: F,
+    ) -> Result<(), Error> {
+        layouter.assign_region(
+            || "keccak table",
+            |mut region| {
+                let mut offset = 0;
+
+                // All zero row to allow simulating a disabled lookup.
+                self.keccak_assign_row(&mut region, offset, F::zero(), F::zero(), 0, F::zero())?;
+                offset += 1;
+
+                for aux in &auxs {
+                    let KeccakAux { input, output } = aux;
+                    let input_rlc =
+                        RandomLinearCombination::random_linear_combine(*input, randomness);
+                    let output_rlc = Word::random_linear_combine(*output, randomness);
+                    self.keccak_assign_row(
+                        &mut region,
+                        offset,
+                        F::one(),
+                        input_rlc,
+                        input.len(),
+                        output_rlc,
+                    )?;
+                    offset += 1;
+                }
+                Ok(())
+            },
+        )?;
+        Ok(())
+    }
+
+    pub fn ecc_chip_config(&self) -> EccConfig {
+        EccConfig::new(self.range_config.clone(), self.main_gate_config.clone())
+    }
+
+    pub fn integer_chip_config(&self) -> IntegerConfig {
+        IntegerConfig::new(self.range_config.clone(), self.main_gate_config.clone())
+    }
+}
+
+pub struct AssignedECDSA<F: FieldExt> {
+    pk_x_le: [AssignedValue<F>; 32],
+    pk_y_le: [AssignedValue<F>; 32],
+    msg_hash_le: [AssignedValue<F>; 32],
+}
+
+#[derive(Debug)]
+pub(crate) struct AssignedSignatureVerify<F: FieldExt> {
+    pub(crate) address: AssignedCell<F, F>,
+    pub(crate) msg_hash_rlc: AssignedCell<F, F>,
+}
+
+// Returns assigned constants [256^1, 256^2, .., 256^{n-1}]
+fn assign_pows_256<F: FieldExt>(
+    ctx: &mut RegionCtx<'_, '_, F>,
+    main_gate: &MainGate<F>,
+    n: usize,
+) -> Result<Vec<AssignedValue<F>>, Error> {
+    let mut pows = Vec::new();
+    for i in 1..n {
+        pows.push(main_gate.assign_constant(ctx, F::from(256).pow(&[i as u64, 0, 0, 0]))?);
+    }
+    Ok(pows)
```

Comment:
Oh but this is not creating any expression.  This is just using the `halo2wrong` `main_gate` to load constants into the `main_gate` chip.

Actually this is not very optimal because I'm loading the 8 powers of 256 for signature verification; but this will be removed in the future and will be replaced by a more efficient WrongInteger to Bytes chip [sneak peak here](https://github.com/ed255/halo2wrong/blob/feature/mul-const/maingate/src/to_bytes.rs)

Reply:
I like it!!!!! :open_mouth: